internal/controller/controllers/agentserviceconfig_controller.go

```diff
@@ -99,30 +99,223 @@ var (
+				Scheme:       scheme,
+				NodeSelector: nodeSelector,
+				Tolerations:  tolerations,
+			},
+		},
+	}
+	ir.components = []component{
+		{"ImageServiceService", aiv1beta1.ReasonImageHandlerServiceFailure, ir.newImageServiceService},
+		{"ImageServiceRoute", aiv1beta1.ReasonImageHandlerRouteFailure, ir.newImageServiceRoute},
+		{"ImageServiceServiceAccount", aiv1beta1.ReasonImageHandlerServiceAccountFailure, ir.newImageServiceServiceAccount},
+		{"ImageServiceConfigMap", aiv1beta1.ReasonConfigFailure, ir.newImageServiceConfigMap},
+	}
+	return ir
+}
+
+func newWebHookComponentReconciler(log logrus.FieldLogger,
+	client client.Client, scheme *runtime.Scheme,
+	nodeSelector map[string]string, tolerations []corev1.Toleration) *webHookComponentReconciler {
+	wr := &webHookComponentReconciler{
+		StandardComponentsReconciler: StandardComponentsReconciler{
+			configReconciler: configReconciler{
+				Log:          log,
+				Client:       client,
+				Scheme:       scheme,
+				NodeSelector: nodeSelector,
+				Tolerations:  tolerations,
+			},
+		},
+	}
+	wr.components = []component{
+		{"AgentClusterInstallValidatingWebHook", aiv1beta1.ReasonValidatingWebHookFailure, wr.newACIWebHook},
+		{"AgentClusterInstallMutatingWebHook", aiv1beta1.ReasonMutatingWebHookFailure, wr.newACIMutatWebHook},
+		{"InfraEnvValidatingWebHook", aiv1beta1.ReasonValidatingWebHookFailure, wr.newInfraEnvWebHook},
+		{"AgentValidatingWebHook", aiv1beta1.ReasonValidatingWebHookFailure, wr.newAgentWebHook},
+		{"WebHookService", aiv1beta1.ReasonWebHookServiceFailure, wr.newWebHookService},
+		{"WebHookServiceDeployment", aiv1beta1.ReasonWebHookDeploymentFailure, wr.newWebHookDeployment},
+		{"WebHookServiceAccount", aiv1beta1.ReasonWebHookServiceAccountFailure, wr.newWebHookServiceAccount},
+		{"WebHookClusterRole", aiv1beta1.ReasonWebHookClusterRoleFailure, wr.newWebHookClusterRole},
+		{"WebHookClusterRoleBinding", aiv1beta1.ReasonWebHookClusterRoleBindingFailure, wr.newWebHookClusterRoleBinding},
+		{"WebHookAPIService", aiv1beta1.ReasonWebHookAPIServiceFailure, wr.newWebHookAPIService},
+	}
+	return wr
+}
+
+func NewAgentServiceConfigReconciler(log logrus.FieldLogger, client client.Client, scheme *runtime.Scheme,
+	nodeSelector map[string]string, tolerations []corev1.Toleration,
+	eventRecorder record.EventRecorder, namespace string) *AgentServiceConfigReconciler {
+	ascr := &AgentServiceConfigReconciler{
+		configReconciler: configReconciler{
+			Log:          log,
+			Client:       client,
+			Scheme:       scheme,
+			NodeSelector: nodeSelector,
+			Tolerations:  tolerations,
+		},
+		Recorder:  eventRecorder,
+		Namespace: namespace,
+	}
+	ascr.subsystemReconcilers = []subsystemReconciler{
+		newAssistedServiceComponentReconciler(log, client, scheme, nodeSelector, tolerations),
+		newImageServiceComponentReconciler(log, client, scheme, nodeSelector, tolerations),
+		newWebHookComponentReconciler(log, client, scheme, nodeSelector, tolerations),
+	}
+	return ascr
+}
```

Comment:
I will do that but some types need to be preserved because aside from breaking the component map I need to move the individual  component reconciliation functions from ASC Reconciler to someplace that can be shared by the HASC Reconciler as well. 

The component level functions need to use some common state (like client and schemes).  If you want to make them global we need to pass scheme and client as parameters to the component level reconcile functions.

Do you have an example for a composite operator that is well built in this regard?

Reply:
Done.